File: 394.字符串解码.py
#
# @lc app=leetcode.cn id=394 lang=python3
#
# [394] 字符串解码
#

# @lc code=start
class Solution:
    def decodeString(self, s: str) -> str:
        stack = []
        numStr = ""
        for ch in s:
            if ch.isdigit():
                numStr = numStr + ch
            elif ch == '[':
                stack.append(numStr)
                numStr = ""
                stack.append(ch)
            elif ch == ']':
                tmpStr = ""
                while stack[-1]!= '[':
                    tmpStr = stack.pop() + tmpStr

                stack.pop()
                num = int(stack.pop())
                stack.append(tmpStr*num)
            else:
                stack.append(ch)
        return "".join(stack)
# @lc code=end

File: 394.字符串解码.java
/*
 * @lc app=leetcode.cn id=394 lang=java
 *
 * [394] 字符串解码
 */

// @lc code=start
class Solution {
    public String decodeString(String s) {
        LinkedList<String> stack = new LinkedList<>();
        StringBuilder numStr = new StringBuilder();

        for (char c : s.toCharArray()) {
            if (c <= '9' && c >= '0') {
                numStr = numStr.append(c);
            } else if (c == '[') {
                stack.add(numStr.toString());
                numStr.setLength(0);
                stack.add("[");

            } else if (c == ']') {
                StringBuilder tmp = new StringBuilder();
                while(stack.peekLast()!="[") {
                    tmp.insert(0, stack.removeLast());
                }

                stack.removeLast();
                int num = Integer.parseInt(stack.removeLast());
                while(num-- > 0) {
                    stack.add(tmp.toString());
                }
            } else {
                stack.add(String.valueOf(c));
            }
        }
        return String.join("", stack);
    }
}
// @lc code=end

